# ĐỒ ÁN CUỐI KỲ MÔN LẬP TRÌNH CHO KHOA HỌC DỮ LIỆU

*Giảng viên: Thầy Trần Trung Kiên*

---

**Nhóm sinh viên thực hiện:**

Sinh viên 01: Trương Trọng Lộc - 18120197

Sinh viên 02: Dương Nam Hải - 19120068

---

## Import các thư viện cần thiết

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

---

## Thu thập dữ liệu

### Tổng quan về dữ liệu

- Tập dữ liệu được sử dụng trong đồ án là tập dữ liệu về các **nhân tố có thể tác động đến tuổi thọ** con người 
(chẳng hạn về giáo dục, y tế, kinh tế) của 183 quốc gia trên thế giới từ năm 2000-2016. Tập dữ liệu tên là **WHO national life expectancy**.
- Tập dữ liệu được lấy và download từ Kaggle thông qua [đường dẫn](https://www.kaggle.com/mmattson/who-national-life-expectancy), bộ dữ liệu do tác giả MMattson thu thập.

### Dữ liệu được phép sử dụng hay không?

License của tập dữ liệu này là [CC0 - Public Domain](https://creativecommons.org/publicdomain/zero/1.0/). Với license này thì ta có thể sử dụng tập dữ liệu mà không cần xin phép tác giả. Cụ thể, license này thể hiện tác giả đã từ bỏ quyền sở hữu của bản thân để trao quyền cho người khác có thể sử dụng để chỉnh sửa, phân phối,... dưới bất kỳ hình thức nào trong phạm vi pháp luật (có thể là học tập, nghiên cứu, thương mại,...).

### Dữ liệu được thu thập như thế nào?

* Dữ liệu được thu thập từ dữ liệu của Tổ chức Y tế Thế giới (World Health Organization) thông qua cả 3 cách sau: 
    * Trang web dữ liệu Thống kê về Giáo dục các Quốc gia của Viện thống kê UNESCO ([UNESCO Databases of Resources on Education](https://en.unesco.org/themes/education/databases));
    * API từ [GHO (Global Health Observatory)](https://www.who.int/data/gho/);
    * Thủ công.
* Chi tiết về cách thu thập dữ liệu có thể tham khảo tại notebook của tác giả tại [đây](https://www.kaggle.com/mmattson/template-to-generate-the-dataset).

---

## Khám phá dữ liệu

### Đọc dữ liệu từ file CSV vào dataframe

Dữ liệu được lưu trong 1 file CSV duy nhất `who_life_exp.csv` và được chứa trong thư mục Data. Ta sử dụng thư viện Pandas để đọc dữ liệu từ CSV này vào Data Frame.

In [2]:
life_exp_df = pd.read_csv('./Data/who_life_exp.csv')
life_exp_df.head()

country country_code  region  year  life_expect  life_exp60  \
0  Angola          AGO  Africa  2000     47.33730    14.73400   
1  Angola          AGO  Africa  2001     48.19789    14.95963   
2  Angola          AGO  Africa  2002     49.42569    15.20010   
3  Angola          AGO  Africa  2003     50.50266    15.39144   
4  Angola          AGO  Africa  2004     51.52863    15.56860   

   adult_mortality  infant_mort  age1-4mort  alcohol  ...  che_gdp    une_pop  \
0         383.5583     0.137985    0.025695  1.47439  ...  1.90860  16395.473   
1         372.3876     0.133675    0.024500  1.94025  ...  4.48352  16945.753   
2         354.5147     0.128320    0.023260  2.07512  ...  3.32946  17519.417   
3         343.2169     0.122040    0.021925  2.20275  ...  3.54797  18121.479   
4         333.8711     0.115700    0.020545  2.41274  ...  3.96720  18758.145   

   une_infant  une_life  une_hiv  une_gni  une_poverty  une_edu_spend  \
0       122.2    46.522      1.0   2530.0         32.3        2.60753   
1       118.9    47.059      1.1   2630.0          NaN            NaN   
2       115.1    47.702      1.2   3180.0          NaN            NaN   
3       110.8    48.440      1.3   3260.0          NaN            NaN   
4       106.2    49.263      1.3   3560.0          NaN            NaN   

   une_literacy  une_school  
0           NaN         NaN  
1      67.40542         NaN  
2           NaN         NaN  
3           NaN         NaN  
4           NaN         NaN  

[5 rows x 32 columns]

### Dữ liệu gồm có bao nhiêu dòng và bao nhiêu cột?

In [3]:
num_rows, num_cols = life_exp_df.shape
num_rows, num_cols

(3111, 32)

**Nhận xét:** 
* Tập dữ liệu gồm 3111 dòng và 32 cột.
* Ngoài ra, theo thông tin của bộ dữ liệu, dữ liệu được khảo sát từ năm 2000-2016 ở 183 quốc gia, điều này hoàn toàn phù hợp với kết quả là 3111 dòng (nói cách dễ hiểu là do 2000-2016 là 17 năm, ta lấy 17 năm * 183 quốc gia = 3111 dòng, dự đoán ban đầu của em là vậy ^^ Em sẽ khảo sát ở phần sau xem thử có đúng như dự đoán không).

### Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?

* Theo quan sát, mỗi dòng thể hiện thông tin về các yếu tố có thể ảnh hưởng đến tuổi thọ (tỉ lệ tiêm chủng, dân số, tỉ lệ người lớn biết chữ, thu nhập bình quân,...) của **một quốc gia** trong **một năm**. 
* Có vẻ **không** có dòng nào có ý nghĩa khác nhau. 

### Dữ liệu có các dòng bị lặp không?

Kiểm tra việc này và lưu kết quả vào biến `have_duplicated_rows`. Biến này sẽ có giá trị True nếu dữ liệu có các dòng bị lặp và có giá trị False nếu ngược lại.

In [4]:
life_exp_dup = life_exp_df.duplicated()
have_duplicated_rows = np.sum(life_exp_dup) != 0
have_duplicated_rows

False

**Nhận xét:** Tập dữ liệu không có dòng nào trùng nhau.

### Mỗi cột có ý nghĩa gì?

Theo mô tả trên Kaggle, mỗi cột có ý nghĩa như sau:

|Tên cột dữ liệu | Mô tả |
|:------:| ---|
|Country| Tên quốc gia khảo sát |
|country_code| Mã định danh quốc gia (3 chữ cái)|
|region| Khu vực của quốc gia khảo sát|
|year| Năm khảo sát|
|life_expect| Tuổi thọ kì vọng khi sinh (đơn vị năm) |
|life_exp60| Tuổi thọ kì vọng khi 60 tuổi (đơn vị năm)|
|adult_mortality|Tỷ lệ tử vong của người trưởng thành (xác suất tử vong từ 15 đến 60 tuổi trên 1000 dân số)|
|infant_mort|Tỷ lệ tử vong dưới 1 tuổi|
|age1-4mort|Tỷ lệ tử vong từ 1 đến 4 tuổi|
|alcohol|Mức tiêu thụ rượu bình quân đầu người từ 15 tuổi trở lên (tính bằng rượu nguyên chất)|
|bmi|BMI trung bình (từ 18 tuổi trở lên) (ước tính chuẩn hóa theo độ tuổi)|
|age5-19thinness|Tỷ lệ suy dinh dưỡng ở trẻ em và thanh thiếu niên (%)|
|age5-19obesity	|Tỷ lệ béo phì ở trẻ em và thanh thiếu niên (%)|
|hepatitis|Tỷ lệ tiêm ngừa viêm gan siêu vi B (HepB) ở trẻ 1 tuổi (%)|
|measles|Tỷ lệ tiêm ngừa sởi mũi đầu tiên (MCV1) ở trẻ 1 tuổi (%)|
|polio| Tỷ lệ tiêm ngừa bại liệt (Pol3) ở trẻ 1 tuổi (%)|
|diphtheria|Tỷ lệ tiêm ngừa bạch hầu, uốn ván và ho gà (DTP3) cho trẻ 1 tuổi (%)|
|basic_water|Tỷ lệ dân số được tiếp cận các dịch vụ nước cơ bản|
|doctors|Mật độ bác sĩ y khoa trên 10.000 dân|
|hospitals|Mật độ bệnh viện trên 100.000 dân|
|gni_capita|Tổng thu nhập quốc dân trên đầu người, đơn vị $ (tổng thu thập mà người dân và doanh nghiệp của quốc gia kiếm được)|
|gghe-d	|Chi tiêu từ ngân sách chính phủ cho lĩnh vực y tế tính theo tỷ lệ phần trăm của tổng sản phẩm quốc nội (GDP) (%)|
|che_gdp|Chi phí người dân bỏ ra cho dịch vụ y tế (CHE) tính theo phần trăm tổng sản phẩm quốc nội (GDP) (%)|
|une_pop|Dân số (đơn vị nghìn)|
|une_infant|Tỷ lệ tử vong ở trẻ sơ sinh trên 1.000 ca sinh sống)|
|une_life|Tuổi thọ kì vọng nếu các hình thức tử vong phổ biến tại thời điểm được sinh ra không thay đổi trong suốt cuộc đời của trẻ|
|une_hiv|Tỷ lệ nhiễm HIV, đơn vị \% trên tổng dân số từ 15-49 tuổi|
|une_gni|Thu nhập quốc dân trên đầu người|
|une_poverty|Tỷ lệ số người nghèo có mức sống 1,90 đô la một ngày (PPP) (\% trên dân số)|
|une_edu_spend|Chi tiêu của chính phủ cho giáo dục tính theo phần trăm GDP (\%)|
|une_literacy|Tỷ lệ người lớn biết chữ, cả hai giới (dân số trên 15 tuổi) (\%)|
|une_school|Số năm đi học trung bình (ISCED 1 trở lên), cả hai giới (dân số trên 25 tuổi)|

**Các thuộc tính có tên `une_`** được thống kê bởi tổ chức UNESCO.

**Các thuộc tính còn lại** được thống kê bởi GHO (Global Health Observatory).


### Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu dữ liệu chưa phù hợp để có thể xử lý tiếp không?

In [5]:
col_dtypes = life_exp_df.dtypes
col_dtypes

country             object
country_code        object
region              object
year                 int64
life_expect        float64
life_exp60         float64
adult_mortality    float64
infant_mort        float64
age1-4mort         float64
alcohol            float64
bmi                float64
age5-19thinness    float64
age5-19obesity     float64
hepatitis          float64
measles            float64
polio              float64
diphtheria         float64
basic_water        float64
doctors            float64
hospitals          float64
gni_capita         float64
gghe-d             float64
che_gdp            float64
une_pop            float64
une_infant         float64
une_life           float64
une_hiv            float64
une_gni            float64
une_poverty        float64
une_edu_spend      float64
une_literacy       float64
une_school         float64
dtype: object

**Nhận xét:** Các cột có kiểu dữ liệu không phải `object` phù hợp, cần làm rõ hơn 3 cột có kiểu dữ liệu đang là `object`, trong Python kiểu dữ liệu `object` thường chỉ chuỗi, nhưng thật chất nó có thể chứa đối tượng bất kỳ vì bên dưới lưu dưới dạng địa chỉ. Trong Data Frame, nếu một cột có `dtype` là `object` thì có thể các phần tử trong cột này sẽ có kiểu dữ liệu khác nhau. Do đó, cần khảo sát từng cột có `dtype` là `object`.

In [6]:
# Hàm trả về các kiểu dữ liệu trong 1 cột
def open_object_dtype(s):
    dtypes = set()
    
    dtypes = set(s.apply(type))
    
    return dtypes

In [7]:
print(open_object_dtype(life_exp_df['country']))
print(open_object_dtype(life_exp_df['country_code']))
print(open_object_dtype(life_exp_df['region']))

{<class 'str'>}
{<class 'str'>}
{<class 'str'>}


Như vậy, cả 3 cột đều có 1 kiểu dữ liệu duy nhất là `str`. 
Do đó **tất cả các cột** trong tập dữ liệu đều đang có kiểu dữ liệu **phù hợp**.

### Với mỗi cột có kiểu dữ liệu dạng numeric, các giá trị được phân bố như thế nào?

* Trong 32 cột dữ liệu của dataset, theo khám phá ở trên có 3 cột là `country`, `country_code` và `region` dạng `object` (cụ thể hơn là `str`). Do đó, ta có tổng cộng 29 cột có kiểu dữ liệu dạng numeric (bài này là `in64` và `float64`). 
* Với mỗi cột numeric, tính tỉ lệ % giá trị thiếu (từ 0 đến 100), min, max. Kết quả được vào Data Frame `nume_col_df`. Dataframe này có 3 dòng là "missing_ratio", "min", "max"; và có 29 cột tương ứng với các cột có `dtype` là `int64` hoặc `float64`.  

In [8]:
def missing_ratio(s):
    return s.isna().mean() * 100
missing_df = life_exp_df.iloc[:,3:].agg([missing_ratio])
nume_col_df = pd.concat([missing_df, life_exp_df.iloc[:,3:].describe()])
nume_col_df

year  life_expect   life_exp60  adult_mortality  \
missing_ratio     0.000000     0.000000     0.000000         0.000000   
count          3111.000000  3111.000000  3111.000000      3111.000000   
mean           2008.000000    69.146384    18.912060       193.527786   
std               4.899767     9.129761     2.885412       114.007447   
min            2000.000000    36.227360    10.728000        49.201850   
25%            2004.000000    63.200095    16.619485       108.269750   
50%            2008.000000    71.597200    18.508260       164.822200   
75%            2012.000000    75.537030    21.099175       250.832300   
max            2016.000000    84.166160    26.394020       696.874200   

               infant_mort   age1-4mort      alcohol          bmi  \
missing_ratio     0.000000     0.000000     1.607200     1.092896   
count          3111.000000  3111.000000  3061.000000  3077.000000   
mean              0.032496     0.003489     4.834781    25.052714   
std               0.030361     0.005146     3.998741     2.193445   
min               0.001470     0.000065     0.000000    19.800000   
25%               0.008255     0.000355     1.198030    23.300000   
50%               0.019995     0.000895     3.993560    25.500000   
75%               0.051720     0.004877     7.723420    26.500000   
max               0.164515     0.039095    20.182460    32.200000   

               age5-19thinness  age5-19obesity  ...      che_gdp  \
missing_ratio         1.092896        1.092896  ...     3.760849   
count              3077.000000     3077.000000  ...  2994.000000   
mean                  5.312447        5.972278  ...     6.110353   
std                   4.652408        4.530812  ...     2.505267   
min                   0.100000        0.100000  ...     1.025160   
25%                   1.800000        2.000000  ...     4.238798   
50%                   3.800000        5.200000  ...     5.758030   
75%                   7.800000        8.900000  ...     7.850327   
max                  28.100000       26.700000  ...    20.413410   

                    une_pop   une_infant     une_life      une_hiv  \
missing_ratio  1.189328e+00     0.000000     0.000000    23.818708   
count          3.074000e+03  3111.000000  3111.000000  2370.000000   
mean           3.707550e+04    30.488428    68.957352     2.038397   
std            1.378377e+05    27.756930     9.359194     4.593804   
min            7.601600e+01     1.600000    39.441000     0.100000   
25%            2.195105e+03     8.000000    62.844500     0.100000   
50%            8.544297e+03    19.500000    71.409000     0.400000   
75%            2.509552e+04    48.050000    75.570645     1.500000   
max            1.414049e+06   142.400000    83.984880    28.200000   

                     une_gni  une_poverty  une_edu_spend  une_literacy  \
missing_ratio       3.760849    70.652523      41.337191     81.645773   
count            2994.000000   913.000000    1825.000000    571.000000   
mean            14964.832999    10.845893       4.532930     81.984472   
std             17495.137508    17.168798       1.754280     19.665588   
min               420.000000     0.100000       0.787440     14.376040   
25%              2970.000000     0.600000       3.262830     72.701130   
50%              8340.000000     3.100000       4.425410     90.953740   
75%             20482.500000    12.400000       5.494980     95.786975   
max            122670.000000    94.100000      14.059080     99.998190   

               une_school  
missing_ratio   74.124076  
count          805.000000  
mean             9.712214  
std              2.854723  
min              0.559340  
25%              7.735940  
50%             10.270400  
75%             12.070610  
max             14.378800  

[9 rows x 29 columns]

In [9]:
missing_df = missing_df.squeeze(axis = 0)
cols_to_drop = missing_df[missing_df > 50]
cols_to_drop

hospitals       95.821279
une_poverty     70.652523
une_literacy    81.645773
une_school      74.124076
Name: missing_ratio, dtype: float64

**Nhận xét:** Có 4 cột là `hospitals`, `une_poverty`, `une_literacy` và `une_school` có tỷ lệ giá trị thiếu lớn hơn 50%. Do thiếu hơn 50% rất khó để khai thác được dữ liệu nên quyết định xóa các cột này.

In [10]:
life_exp_df = life_exp_df.drop(cols_to_drop.index, 1)
life_exp_df

country country_code           region  year  \
0                               Angola          AGO           Africa  2000   
1                               Angola          AGO           Africa  2001   
2                               Angola          AGO           Africa  2002   
3                               Angola          AGO           Africa  2003   
4                               Angola          AGO           Africa  2004   
...                                ...          ...              ...   ...   
3106  Lao People's Democratic Republic          LAO  Western Pacific  2012   
3107  Lao People's Democratic Republic          LAO  Western Pacific  2013   
3108  Lao People's Democratic Republic          LAO  Western Pacific  2014   
3109  Lao People's Democratic Republic          LAO  Western Pacific  2015   
3110  Lao People's Democratic Republic          LAO  Western Pacific  2016   

      life_expect  life_exp60  adult_mortality  infant_mort  age1-4mort  \
0        47.33730    14.73400         383.5583     0.137985    0.025695   
1        48.19789    14.95963         372.3876     0.133675    0.024500   
2        49.42569    15.20010         354.5147     0.128320    0.023260   
3        50.50266    15.39144         343.2169     0.122040    0.021925   
4        51.52863    15.56860         333.8711     0.115700    0.020545   
...           ...         ...              ...          ...         ...   
3106     64.49448    16.57977         209.2666     0.055840    0.004845   
3107     64.79716    16.64219         205.1459     0.055350    0.004600   
3108     65.10757    16.70787         200.8204     0.054585    0.004405   
3109     65.45008    16.77074         196.6613     0.053465    0.004160   
3110     65.79089    16.83024         192.7939     0.051895    0.003990   

      alcohol  ...  doctors  gni_capita   gghe-d  che_gdp    une_pop  \
0     1.47439  ...      NaN      2190.0  1.11099  1.90860  16395.473   
1     1.94025  ...      NaN      2290.0  2.04631  4.48352  16945.753   
2     2.07512  ...      NaN      2690.0  1.30863  3.32946  17519.417   
3     2.20275  ...      NaN      2820.0  1.46560  3.54797  18121.479   
4     2.41274  ...    0.621      3080.0  1.68663  3.96720  18758.145   
...       ...  ...      ...         ...      ...      ...        ...   
3106  6.29833  ...    1.800      4170.0  0.43670  2.07638   6444.530   
3107  6.42186  ...    4.493      4570.0  0.71266  2.39985   6541.304   
3108  6.64447  ...    4.949         NaN  0.68373  2.29846   6639.756   
3109  6.83826  ...      NaN         NaN  0.86391  2.45366   6741.164   
3110  7.11148  ...      NaN         NaN  0.76495  2.36087   6845.846   

      une_infant  une_life  une_hiv  une_gni  une_edu_spend  
0          122.2    46.522      1.0   2530.0        2.60753  
1          118.9    47.059      1.1   2630.0            NaN  
2          115.1    47.702      1.2   3180.0            NaN  
3          110.8    48.440      1.3   3260.0            NaN  
4          106.2    49.263      1.3   3560.0            NaN  
...          ...       ...      ...      ...            ...  
3106        47.5    65.267      0.3   4570.0        1.82147  
3107        45.6    65.717      0.3   4980.0        3.23381  
3108        43.7    66.144      0.3   5440.0        2.93781  
3109        42.0    66.546      0.3   5810.0            NaN  
3110        40.4    66.924      0.3   6190.0            NaN  

[3111 rows x 28 columns]

### Với mỗi cột có kiểu dữ liệu dạng categorical, các giá trị được phân bố như thế nào?

* Theo khám phá ở trên có 3 cột là `country`, `country_code` và `region` dạng `object` (cụ thể hơn là `str`). Do đó, kiểu dữ liệu dạng categorical của tập dữ liệu này gồm 3 cột đẫ nêu. 
* Với mỗi cột dạng categorical, tính tỉ lệ % giá trị thiếu (từ 0 đến 100), số lượng giá trị khác nhau (không xét giá trị thiếu), list/array các giá trị khác nhau (không xét giá trị thiếu). Kết quả được lưu vào Data Frame `cate_col_df`. Dataframe này có 3 dòng là "missing_ratio", "num_diff_vals", "diff_vals"; và có 3 cột là 'country', 'country_code', 'region'. 

In [11]:
def unique_not_miss(s):
    return s.dropna().unique()
cate_col_df = life_exp_df.iloc[:,:3] \
                            .agg([missing_ratio, pd.Series.nunique, unique_not_miss])

cate_col_df.index = ['missing_ratio','num_diff_vals','diff_vals']
cate_col_df

country  \
missing_ratio                                                0.0   
num_diff_vals                                                183   
diff_vals      [Angola, Burundi, Benin, Burkina Faso, Botswan...   

                                                    country_code  \
missing_ratio                                                0.0   
num_diff_vals                                                183   
diff_vals      [AGO, BDI, BEN, BFA, BWA, CAF, TZA, UGA, ZAF, ...   

                                                          region  
missing_ratio                                                0.0  
num_diff_vals                                                  6  
diff_vals      [Africa, Americas, Eastern Mediterranean, Euro...

**Nhận xét:** Cả 3 cột dạng categorical đều không có giá trị thiếu. Hai cột `country` và `country_code` có 183 giá trị khác nhau, riêng cột `region` có 6 giá trị khác nhau. Điều này hoàn toàn phù hợp vì có 183 được khảo sát trong tập dữ liệu này, các quốc gia khác nhau có thể thuộc chung 1 khu vực.

---

## Đưa ra câu hỏi cần trả lời

### Tiền xử lý

### Phân tích dữ liệu để trả lời cho các câu hỏi 

---

## Nhìn lại quá trình làm đồ án

---

## Tài liệu tham khảo

---